eureka2-client/src/main/java/com/netflix/eureka2/client/resolver/RoundRobinServerResolver.java

```diff
@@ -0,0 +1,143 @@
+ * not emit any elements until the resolver have warmed up for the first time, i.e. the resolver has been loaded
+ * with an initial list of servers. The warm up time can be changed from defaults by calling
+ * {@link #withWarmUpConfiguration(int, TimeUnit)}.
+ *
+ * @author David Liu
+ */
+public class RoundRobinServerResolver implements ServerResolver {
+
+    private static final Logger logger = LoggerFactory.getLogger(RoundRobinServerResolver.class);
+
+    private static final Exception SERVER_CACHE_EMPTY_EXCEPTION = new NoSuchElementException("No servers available for this resolver");
+
+    private final int warmUpTimeout;
+    private final TimeUnit timeUnit;
+
+    private final Observable<ChangeNotification<Server>> serverSource;
+    private final AtomicReference<List<Server>> serverCacheRef;
+    private final AtomicInteger positionRef;
+
+    protected RoundRobinServerResolver(Server... servers) {
+        this(sourceFromList(servers), 10, TimeUnit.SECONDS);
+    }
+
+    protected RoundRobinServerResolver(Observable<ChangeNotification<Server>> serverSource) {
+        this(serverSource, 10, TimeUnit.SECONDS);
+    }
+
+    protected RoundRobinServerResolver(Observable<ChangeNotification<Server>> serverSource, int warmUpTimeout, TimeUnit timeUnit) {
+        this.serverSource = serverSource;
+        this.warmUpTimeout = warmUpTimeout;
+        this.timeUnit = timeUnit;
+        this.serverCacheRef = new AtomicReference<List<Server>>(new ArrayList<Server>());
+        this.positionRef = new AtomicInteger(new Random().nextInt(100));
+    }
+
+    public RoundRobinServerResolver withWarmUpConfiguration(int newWarmUpTimeout, TimeUnit newTimeUnit) {
+        return new RoundRobinServerResolver(serverSource, newWarmUpTimeout, newTimeUnit);
+    }
+
+    @Override
+    public void close() {
+        serverCacheRef.get().clear();
+    }
+
+    @Override
+    public Observable<Server> resolve() {
+        return refreshServerCache().concatMap(new Func1<List<Server>, Observable<? extends Server>>() {
+            @Override
+            public Observable<? extends Server> call(List<Server> servers) {
+                if (servers.isEmpty()) {
+                    return Observable.error(SERVER_CACHE_EMPTY_EXCEPTION);
+                }
+
+                int currentPos = positionRef.getAndIncrement();
+                Server toReturn = servers.get(currentPos % servers.size());
+                return Observable.just(toReturn);
+            }
+        });
+    }
+
+    /**
+     * Connect and warm up the serverCache to the server source. onComplete the result Observable once the
+     * serverCache has warmed up.
+     */
+    private Observable<List<Server>> refreshServerCache() {
+        return serverSource
+                .compose(ChangeNotifications.<Server>buffers())
+                .compose(ChangeNotifications.<Server>snapshots())
+                .materialize()
```

Comment:
Recent refactorings in ChangeNotifications allow for empty server lists. I think in the very beginning we should not emit empty list (filter empty lists out), but we should allow them afterwards, which means that all previously advertised server gone away.
The bottom line is that ChangeNotifications transformers must offer enough flexibility to do that.